src/esp/physics/bullet/BulletArticulatedObject.h

```diff
@@ -0,0 +1,401 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#ifndef ESP_PHYSICS_BULLET_BULLETARTICULATEDOBJECT_H_
+#define ESP_PHYSICS_BULLET_BULLETARTICULATEDOBJECT_H_
+
+/** @file
+ * @brief Class @ref esp::physics::BulletArticulatedObject
+ */
+
+#include <utility>
+
+#include "../ArticulatedObject.h"
+#include "BulletBase.h"
+#include "BulletDynamics/Featherstone/btMultiBodyJointMotor.h"
+#include "BulletDynamics/Featherstone/btMultiBodySphericalJointMotor.h"
+#include "BulletURDFImporter.h"
+
+namespace esp {
+
+namespace physics {
+
+// forward delcaration from BulletURDFImporter
+struct JointLimitConstraintInfo;
+
+////////////////////////////////////
+// Link
+////////////////////////////////////
+
+class BulletArticulatedLink : public ArticulatedLink, public BulletBase {
+ public:
+  BulletArticulatedLink(scene::SceneNode* bodyNode,
+                        const assets::ResourceManager& resMgr,
+                        std::shared_ptr<btMultiBodyDynamicsWorld> bWorld,
+                        int index,
+                        std::shared_ptr<std::map<const btCollisionObject*, int>>
+                            collisionObjToObjIds)
+      : ArticulatedLink(bodyNode, index, resMgr),
+        BulletBase(std::move(bWorld), std::move(collisionObjToObjIds)) {}
+
+  Magnum::Range3D getCollisionShapeAabb() const override {
+    // TODO: collision object should be linked here
+    return Magnum::Range3D();
+  }
+
+  //! link can't do this.
+  void setMotionType(CORRADE_UNUSED MotionType mt) override {}
```

Comment:
Speaking to C++ inheritance philosophy, it seems awkward for BulletArticulatedLink to inherit from BulletBase and yet not offer the full BulletBase interface. As some future refactor, I'd like to suggest switching from inheritance to composition (BulletArticulatedLink owns a BulletBase instance as a member but doesn't publicly present itself as a BulletBase object).

Is silently failing the right thing to do here? I'm imaging some future user getting creative and trying to set motion type per link.

Reply:
Added a warning.